### PR TITLE
feat(redeem): Liquidation (burn) redeem

### PR DESCRIPTION
**Pull request by savudani8**
It takes about 15 mins for the staked relayer to report the theft, which makes the integration tests take a while

### CODE REVIEW

**Comment by nud3l** [comment]
> It takes about 15 mins for the staked relayer to report the theft, which makes the integration tests take a while

Do you know if there is a way to "classify" tests easily? I'm thinking about having certain tests tagged as release tests that we only ever want to execute when we tag a new release. Because waiting 15 min on every pipeline is not really something any of us would want.

The straightforward way would of course be to have the tests in a separate folder like the unit and integration tests. But technically this is also an integration test so logically it Fotos be in the same folder.

File: src/parachain/collateral.ts
Comment (nud3l) [inline src/parachain/collateral.ts:18]: I'd suggest we remove the DOT part in the name since this will change on Kusama and this is anyway a breaking change for the UI.

File: package.json
Comment (nud3l) [inline package.json:3]: Should be minor release due to breaking changes

File: src/parachain/redeem.ts
Comment (nud3l) [inline src/parachain/redeem.ts:137]: Actually that is not always the case. In massive dot exchange rate crashes we might end up with lower dot amounts than you get for burning polkabtc

File: src/parachain/redeem.ts
Comment (nud3l) [inline src/parachain/redeem.ts:127]: Seems like a typo? This didn't change the rate but burns polkabtc for dot

File: src/parachain/redeem.ts
Comment (nud3l) [inline src/parachain/redeem.ts:218]: This API was a really nice addition!

File: src/parachain/vaults.ts
Comment (nud3l) [inline src/parachain/vaults.ts:373]: Do you need that API since it does the same as the call below?

Comment (savudani8) [reply src/parachain/vaults.ts:373]: It's more lightweight than the call below, so I thought it might be useful

**Comment by savudani8** [comment]
> > It takes about 15 mins for the staked relayer to report the theft, which makes the integration tests take a while
> 
> Do you know if there is a way to "classify" tests easily? I'm thinking about having certain tests tagged as release tests that we only ever want to execute when we tag a new release. Because waiting 15 min on every pipeline is not really something any of us would want.
> 
> The straightforward way would of course be to have the tests in a separate folder like the unit and integration tests. But technically this is also an integration test so logically it Fotos be in the same folder.

We could do this via the workflow actions. First, we need to suffix test names as a tag, to achieve this separation. In `package.json`:
- "test:integration:staging": "mocha test/integration/**/*.staging.test.ts
- "test:integration:release": "mocha test/integration/**/*.release.test.ts
This means we'll have two versions of [test.yml](https://github.com/interlay/polkabtc-js/blob/master/.github/workflows/test.yml), one triggered by a push to the master branch and one triggered by PRs.
Then, [npm.yml](https://github.com/interlay/polkabtc-js/blob/master/.github/workflows/npm.yml) will depend on `test-master.yml` to succeed. A problem here is that if a tag is pushed without `test-master.yml` being successful, the `npm.yml`  action will have to be restarted manually. This is to avoid needlessly pushing tags to trigger the npm release script.

@theodorebugnet What do you think?

**Comment by nud3l** [comment]
I think one way around that issue is that on successful completion of the release pipeline, the pipeline can just tag the version and thereby trigger the npm pipeline. That would also resolve the manual step of us tagging. I also don't really see a use case wehre we want to update master without a new release.

@ns212 I think that should be possible with jenkins?

We can do the same for stats btw. What do you guys think?

**Comment by nud3l** [comment]
Pending the testing/pipeline change this LGTM.

**Comment by savudani8** [comment]
@nud3l How would the auto-tagging work though? You mentioned a tool a while ago that parses commit messages to do semantic versioning. Perhaps this one: https://github.com/semantic-release/semantic-release

**Comment by savudani8** [comment]
We cannot enforce `release` integration tests as a dependency of the `npm` action, because they are part of different workflow files. So for now, tagging must happen after manually ensuring that the `release` pipeline succeeded. 

I removed the integration test that checks for successful provider disconnection, because the parachain doesn't process it quickly enough, causing it to fail (here is the provider logic https://github.com/polkadot-js/api/blob/master/packages/rpc-provider/src/ws/index.ts#L355). Until now, that test was left out of the `integration` test directory and was never run.

**Comment by nud3l** [comment]
> We cannot enforce `release` integration tests as a dependency of the `npm` action, because they are part of different workflow files. So for now, tagging must happen after manually ensuring that the `release` pipeline succeeded. 
> 
> 
> 
> I removed the integration test that checks for successful provider disconnection, because the parachain doesn't process it quickly enough, causing it to fail (here is the provider logic https://github.com/polkadot-js/api/blob/master/packages/rpc-provider/src/ws/index.ts#L355). Until now, that test was left out of the `integration` test directory and was never run.
> 
> 

Can you check with Nikolai? I think Jenkins allows us to trigger an action that would tag the repo when the pipeline on master succeeds automatically. Until, then we can tag manually.

If we go for the manual version, for now please add a ticket for Nikolai to add the automatic tagging to Jenkins.

File: .github/workflows/test-release.yml
Comment (nud3l) [inline .github/workflows/test-release.yml:20]: This should run all tests on master, so just 'yarn test' I think

File: package.json
Comment (nud3l) [inline package.json:28]: We could leave this to run both? Maybe you'd want to do that locally in a single command

File: src/parachain/collateral.ts
Comment (nud3l) [inline src/parachain/collateral.ts:18]: Thinking about this, would it make more sense to return a Big denominated in DOT actually? We are mostly just handing BTC and DOT amounts in the UI and there is little advantage to have manual denomination conversions

File: .github/workflows/test-release.yml
Comment (savudani8) [reply .github/workflows/test-release.yml:20]: Since the staging tests passed in the PR, I don't think we need to re-run them on `master`

Comment (nud3l) [reply .github/workflows/test-release.yml:20]: following scenario I have in mind:
PR1 changes staging tests X
PR2 changes staging tests Y

Both are passing. But there is an edge case where PR2 affects tests in PR1 that have just changed. Now, independently they pass, but on master they could fail. Example could be when we change the way that issue fees are deducted. In PR1, we could have fixed the issue tests. PR2 could be update of the vault struct. But maybe in the PR2, we added a new test that gets all the vault structs with the issuable tokens of the vault. But now that test would be incorrect because of PR1.

It's best practice to run all tests before a release to cover all of these edge cases.

**Comment by savudani8** [comment]
The APY fix increases the displayed value by ~100x (e.g. from `0.0068` to `0.235` on my local UI)

**Review by nud3l — COMMENTED** [review]
Pending the below: what actually changed in the calculation? Keep in mind that the UI will fetch the exchange rate globally and we could possibly just submit it as a parameter to make sure this call takes less time (i.e. on app load we already have the exchange rate so when we open the vault dashboard, we would not have to wait for this APY call to return the exchange rate promise since we don't have an API call).

It was submitted as a parameter before, but was that part incorrect since the logic itself did not change?

**Comment by savudani8** [comment]
> Keep in mind that the UI will fetch the exchange rate globally and we could possibly just submit it as a parameter to make sure this call takes less time (i.e. on app load we already have the exchange rate so when we open the vault dashboard, we would not have to wait for this APY call to return the exchange rate promise since we don't have an API call).


Good point, I'll leave the `dotToBtcRate` as an optional parameter though, in case the caller does not have it (I suggested a similar approach in the requests aggregator PR, when talking about `UI Flow / Caching` https://github.com/interlay/polkabtc-js/pull/272#issuecomment-806080122). 

> Pending the below: what actually changed in the calculation? 
> It was submitted as a parameter before, but was that part incorrect since the logic itself did not change?

What changed is that previously, we were passing small denominations (sat and planck) but converting using the DOT/BTC rate, whereas now we actually pass the large denominations (hence the `Big`). 

Here is an example of where the call is wrong: https://github.com/interlay/polkabtc-js/blob/master/src/parachain/staked-relayer.ts#L385

The APY fix commit returns Big values instead of BN-like values when fetching client fees: https://github.com/interlay/polkabtc-js/pull/277/commits/d19167912c298b865b5466a9a528e52e8cd45b18#diff-63cf19a974a0d655c3d6dfbe5de1fcbe12c399d0cb8b0dce382a1c04ad30ac4dR379

**Comment by nud3l** [comment]
> > Keep in mind that the UI will fetch the exchange rate globally and we could possibly just submit it as a parameter to make sure this call takes less time (i.e. on app load we already have the exchange rate so when we open the vault dashboard, we would not have to wait for this APY call to return the exchange rate promise since we don't have an API call).
> 
> Good point, I'll leave the `dotToBtcRate` as an optional parameter though, in case the caller does not have it (I suggested a similar approach in the requests aggregator PR, when talking about `UI Flow / Caching` [#272 (comment)](https://github.com/interlay/polkabtc-js/pull/272#issuecomment-806080122)).

Yep, that seems the best approach!

> 
> > Pending the below: what actually changed in the calculation?
> > It was submitted as a parameter before, but was that part incorrect since the logic itself did not change?
> 
> What changed is that previously, we were passing small denominations (sat and planck) but converting using the DOT/BTC rate, whereas now we actually pass the large denominations (hence the `Big`).
> 
> Here is an example of where the call is wrong: https://github.com/interlay/polkabtc-js/blob/master/src/parachain/staked-relayer.ts#L385
> 
> The APY fix commit returns Big values instead of BN-like values when fetching client fees: [d191679#diff-63cf19a974a0d655c3d6dfbe5de1fcbe12c399d0cb8b0dce382a1c04ad30ac4dR379](https://github.com/interlay/polkabtc-js/commit/d19167912c298b865b5466a9a528e52e8cd45b18#diff-63cf19a974a0d655c3d6dfbe5de1fcbe12c399d0cb8b0dce382a1c04ad30ac4dR379)

Ah, right makes sense! Thanks for fixing it!

**Review by nud3l — APPROVED** [review]

**Pull request merged**
